src/util/misc.ts

```diff
@@ -195,56 +233,145 @@ import { cos } from './index.ts';
      * @param {number} options.scaleX
      * @param {number} options.scaleY
      * @param {boolean} [openPath] whether the shape is open or not, affects the calculations of the first and last points
-     * @returns {fabric.Point[]} array of size 2n/4n of all suspected points
+     * @param {boolean} [traceProjections] also returns the point and bisector that originated the projection
+     * @returns {fabric.Point[]} array of size n (for miter stroke) or 2n (for bevel or round) of all suspected points
      */
-    projectStrokeOnPoints: function (points, options, openPath) {
-      var coords = [], s = options.strokeWidth / 2,
+    projectStrokeOnPoints: function (points, options, openPath, traceProjections /** = false */) {
+      //  TODO remove next line after es6 migration and set `counterClockwise` arg to `true` by default
+      traceProjections = traceProjections === true;
+
+      if (points.length <= 1) { return []; }
+
+      var coords = [],
+          s = options.strokeWidth / 2,
+          scale = new fabric.Point(options.scaleX, options.scaleY),
           strokeUniformScalar = options.strokeUniform ?
-            new fabric.Point(1 / options.scaleX, 1 / options.scaleY) : new fabric.Point(1, 1),
-          getStrokeHatVector = function (v) {
-            var scalar = s / (Math.hypot(v.x, v.y));
-            return new fabric.Point(v.x * scalar * strokeUniformScalar.x, v.y * scalar * strokeUniformScalar.y);
-          };
-      if (points.length <= 1) {return coords;}
+            new fabric.Point(1 / options.scaleX, 1 / options.scaleY) :
+            new fabric.Point(1, 1);
+
+      function scaleHatVector(hatVector, scalar) {
+        return hatVector.multiply(strokeUniformScalar).scalarMultiply(scalar);
+      }
+
+      function storeTraceProjections(projection, origin, bisector) {
```

Comment:
I don't mind returning more context
but let's do it always without the `traceProjection` flag
Also, point will become a class today so maybe we return an object instead of mutating it (TS will not like that and I agree)
```ts
{ projectedPoint, originPoint, miterVector, bisectorVector, ...context }
```